sdk/lusid/utilities/api_configuration_loader.py

```diff
@@ -1,55 +1,69 @@
+        def vars_from_env():
+            # Load our configuration details from the environment variables
+            return {
+                "token_url": os.getenv("FBN_TOKEN_URL", None),
+                "api_url": os.getenv("FBN_LUSID_API_URL", None),
+                "username": os.getenv("FBN_USERNAME", None),
+                "password": os.getenv("FBN_PASSWORD", None),
+                "client_id": os.getenv("FBN_CLIENT_ID", None),
+                "client_secret": os.getenv("FBN_CLIENT_SECRET", None),
+                "app_name": os.getenv("FBN_APP_NAME", "")
+            }
+
+        env_vars = vars_from_env()
+        missing = {k: v for k, v in env_vars.items() if v is None}
+        missing.pop("app_name", None)
 
         # If any of the environmental variables are missing use a local secrets file
-        if token_url is None or username is None or password_raw is None or client_id_raw is None \
-                or client_secret_raw is None or api_url is None:
+        if len(missing) > 0:
+
+            if api_secrets_filename is None:
+                var_to_env = {
+                    "token_url": "FBN_TOKEN_URL",
+                    "api_url": "FBN_LUSID_API_URL",
+                    "username": "FBN_USERNAME",
+                    "password": "FBN_PASSWORD",
+                    "client_id": "FBN_CLIENT_ID",
+                    "client_secret": "FBN_CLIENT_SECRET"
+                }
+
+                raise ValueError(f"secrets not specified and missing the following environment variables: {','.join([var_to_env[k] for k in missing.keys()])}")
 
             dir_path = os.path.dirname(os.path.realpath(__file__))
             with open(os.path.join(dir_path, api_secrets_filename), "r") as secrets:
                 config = json.load(secrets)
 
             token_url = os.getenv("FBN_TOKEN_URL", config["api"]["tokenUrl"])
             username = os.getenv("FBN_USERNAME", config["api"]["username"])
-            password = pathname2url(os.getenv("FBN_PASSWORD", config["api"]["password"]))
-            client_id = pathname2url(os.getenv("FBN_CLIENT_ID", config["api"]["clientId"]))
-            client_secret = pathname2url(os.getenv("FBN_CLIENT_SECRET", config["api"]["clientSecret"]))
+            password = os.getenv("FBN_PASSWORD", config["api"]["password"])
+            client_id = os.getenv("FBN_CLIENT_ID", config["api"]["clientId"])
+            client_secret = os.getenv("FBN_CLIENT_SECRET", config["api"]["clientSecret"])
             api_url = os.getenv("FBN_LUSID_API_URL", config["api"]["apiUrl"])
             app_name = os.getenv("FBN_APP_NAME", config["api"].get("applicationName", ""))
 
+            return ApiConfiguration(
+                token_url=token_url,
+                api_url=api_url,
+                username=username,
+                password=password,
+                client_id=client_id,
+                client_secret=client_secret,
+                app_name=app_name
+            )
```

Comment:
What happens if there are still missing values at this point? Looks like it won't fail gracefully.

Reply:
refactored